src/Microsoft.SqlTools.ServiceLayer/QueryExecution/SerializationService.cs

```diff
@@ -40,93 +41,106 @@ public override void InitializeService(IProtocolEndpoint serviceHost)
         /// <summary>
         /// Begin to process request to save a resultSet to a file in CSV format
         /// </summary>
-        internal async Task HandleSerializeStartRequest(SerializeDataStartRequestParams serializeParams,
+        internal Task HandleSerializeStartRequest(SerializeDataStartRequestParams serializeParams,
             RequestContext<SerializeDataResult> requestContext)
+        {
+            // Run in separate thread so that message thread isn't held up by a potentially time consuming file write
+            Task.Run(async () => {
+                await RunSerializeStartRequest(serializeParams, requestContext);
+            }).ContinueWithOnFaulted(async t => await SendErrorAndCleanup(serializeParams?.FilePath, requestContext, t.Exception));
```

Comment:
So a potential problem is that now since this is truly async we'll immediately return  - at which point the caller may then call to continue serialization. But the start request may not actually have started/finished yet and thus we have a race condition if the continue serialization call starts trying to do stuff assuming that the start request has finished.

Same issue with the continue calls - those could then get out of order or start causing problems trying to write at the same time

Reply:
So, this is dependent on the caller waiting on responses. That's what I've implemented in ADS - we wait on the result to come back, then send the next request. Since each one is awaited, we will get correct ordering. Right now I would treat this as "by design" because:
- the cost to implement differently would be a lot higher since you then need an ordered queue model to track
- most importantly since we're sending potentially large amounts of data, all of which are kept in memory until serialized, I explicitly do not want to encourage "send 5 messages in a row" and instead require "send message, wait until processed, send next message".

Reply:
Oh right, yeah my mistake. Got confused about the return value here being the indication that the request was done. Nevermind!

Reply:
I suggest leaving a comment in the code about this stated 'by design' contract.